tokio/src/sync/semaphore_ll.rs

```diff
@@ -501,86 +596,149 @@ impl Permit {
+            _ => false,
+        }
     }
 
     /// Try to acquire the permit. If no permits are available, the current task
     /// is notified once a new permit becomes available.
     pub(crate) fn poll_acquire(
         &mut self,
         cx: &mut Context<'_>,
+        num_permits: u16,
         semaphore: &Semaphore,
     ) -> Poll<Result<(), AcquireError>> {
+        use std::cmp::Ordering::*;
+        use PermitState::*;
+
         match self.state {
-            PermitState::Idle => {}
-            PermitState::Waiting => {
+            Waiting(requested) => {
+                // There must be a waiter
                 let waiter = self.waiter.as_ref().unwrap();
 
-                if waiter.acquire(cx)? {
-                    self.state = PermitState::Acquired;
+                match requested.cmp(&num_permits) {
+                    Less => {
+                        let delta = num_permits - requested;
+
+                        // Request additional permits. If the waiter has been
+                        // dequeued, it must be re-queued.
+                        if !waiter.try_inc_permits_to_acquire(delta as usize) {
+                            let waiter = NonNull::from(&**waiter);
+
+                            // Ignore the result. The check for
+                            // `permits_to_acquire()` will converge the state as
+                            // needed
+                            let _ = semaphore.poll_acquire2(delta, || Some(waiter))?;
+                        }
+
+                        self.state = Waiting(num_permits);
+                    }
+                    Greater => {
+                        let delta = requested - num_permits;
+                        let to_release = waiter.try_dec_permits_to_acquire(delta as usize);
+
+                        semaphore.add_permits(to_release);
+                        self.state = Waiting(num_permits);
+                    }
+                    Equal => {}
+                }
+
+                if waiter.permits_to_acquire()? == 0 {
+                    self.state = Acquired(requested);
+                    return Ready(Ok(()));
+                }
+
+                waiter.waker.register_by_ref(cx.waker());
+
+                if waiter.permits_to_acquire()? == 0 {
+                    self.state = Acquired(requested);
                     return Ready(Ok(()));
-                } else {
-                    return Pending;
                 }
-            }
-            PermitState::Acquired => {
-                return Ready(Ok(()));
-            }
-        }
 
-        match semaphore.poll_permit(Some((cx, self)))? {
-            Ready(()) => {
-                self.state = PermitState::Acquired;
-                Ready(Ok(()))
-            }
-            Pending => {
-                self.state = PermitState::Waiting;
                 Pending
             }
+            Acquired(acquired) => {
+                if acquired >= num_permits {
```

Comment:
The `>` case is a bit confusing. Can it happen if a follow-up poll call uses a different number of permits? Should this release the extra permits? Wondering who is tracking how many extra permits need to be released

Reply:
The `Permit` handle itself tracks the number of acquired permits.

This case probably doesn't happen in practice (especially since this is a private type) but could happen if `poll_acquire` is called in a weird way. Ideally we do a bigger refactor of semaphore in the near future.